Sources/SwiftKuery/ConnectionPool.swift

```diff
@@ -104,13 +104,78 @@ public class ConnectionPool {
         release()
     }
     
+    // Check to see whether the pool's capacity has fallen to zero, and if so, attempt
+    // to grow the pool so there is one item in it.
+    //
+    // Since the pool's capacity would only fall to zero when connectivity problems exist,
+    // it is possible that the generation of a new connection will fail and the pool will
+    // still have zero capacity.
+    //
+    // Returns true iff the pool has a non-zero capacity.
+    private func tryToEnsureNonZeroCapacity() -> Bool {
```

Comment:
We need to check capacity again after we lock. 

We check capacity is equal to zero, if it is we lock the pool. However a connection may have also been returned to the pool in the small window between us checking the capacity and us locking the pool, which would increase the capacity and remove the need for us to grow the pool. 

I discussed this with Dave and suggested adding a comment here to explain what we are doing and why, as it does look strange that we check capacity twice.

Reply:
@ianpartridge You're right, this needs commenting.

We don't want to lock up-front, because locking involves a semaphore (synchronization), and this is on the critical path.  In normal circumstances, the capacity is greater than zero (normal operation) so we don't want to pay a cost for locking unless we have to.

If the first check succeeds, then we (probably) have to do some work, so we lock and then check again.  We have to check again, because multiple threads could simultaneously check for `capacity == 0` outside the lock, and we don't want them all to generate an item.